community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/storage/paged/PagedIndexInput.java

```diff
@@ -0,0 +1,456 @@
+            long startPosition, long size ) throws IOException
+    {
+        this( new InputResources.RootInputResources( pagedFile ),
+                resourceDescription, startPosition, pagedFile.pageSize(),
+                size );
+    }
+
+    PagedIndexInput( InputResources resources, String resourceDescription,
+            long startPosition, int pageSize, long size ) throws IOException
+    {
+        super( resourceDescription );
+
+        this.pageSize = pageSize;
+        this.length = size;
+        this.startPosition = startPosition;
+        this.currentPageId = pageId( 0 );
+        this.currentPageOffset = pageOffset( 0 );
+        this.endPageId = pageId( startPosition + size );
+        this.endPageOffset = lastPageOffset( startPosition + size, pageSize );
+
+        this.resources = resources;
+        this.cursor = resources.openCursor( currentPageId, this );
+    }
+
+    private static int lastPageOffset( long fileSize, int pageSize )
+    {
+        if ( fileSize == 0 )
+        {
+            return 0;
+        }
+        int size = (int) (fileSize % pageSize);
+        return size == 0 ? pageSize : size;
+    }
+
+    private long pageId( long position )
+    {
+        return (startPosition + position) / pageSize;
+    }
+
+    private int pageOffset( long position )
+    {
+        return (int) ((startPosition + position) % pageSize);
+    }
+
+    @Override
+    public final byte readByte() throws IOException
+    {
+        byte val = readByte( currentPageId, currentPageOffset );
+        incrementPageOffset( 1 );
+        return val;
+    }
+
+    @Override
+    public byte readByte( long pos ) throws IOException
+    {
+        return readByte( pageId( pos ), pageOffset( pos ) );
+    }
+
+    private byte readByte( long pageId, int pageOffset ) throws IOException
+    {
+        moveCursor( pageId, pageOffset );
+
+        byte val;
+        do
+        {
+            val = cursor.getByte( pageOffset );
+        }
+        while ( cursor.shouldRetry() );
+
+        boundsCheck();
+        return val;
+    }
+
+    @Override
+    public short readShort() throws IOException
+    {
+        short val = readShort( currentPageId, currentPageOffset );
+        incrementPageOffset( 2 );
+        return val;
+    }
+
+    @Override
+    public short readShort( long pos ) throws IOException
+    {
+        return readShort( pageId( pos ), pageOffset( pos ) );
+    }
+
+    private short readShort( long pageId, int pageOffset ) throws IOException
+    {
+        // Cross-page read?
+        if ( pageSize - pageOffset < 2 )
+        {
+            long originalPageId = currentPageId;
+            int originalPageOffset = currentPageOffset;
+            try
+            {
+                currentPageId = pageId;
+                currentPageOffset = pageOffset;
+                // Delegate to parent, which reads byte-by-byte
+                return super.readShort();
```

Comment:
Hmm.. this is interesting, yeah.. I guess if we actually had concurrent writes, the only way to do this would be to write-lock the two pages? There's currently no way to do `shouldRetry` across two pages right?

That said, yes, this should be safe since we're guaranteed read-only

Reply:
You can do shoudRetry across multiple pages by opening linked page cursors.